### PR TITLE
Fix modal fit on iOS for responsiveness

### PR DESCRIPTION
**Pull request by joaom182**
Related to #957 

### Before
modal-example

### After
Simulator Screen Shot - iPhone X - 2019-03-29 at 15 56 15

### CODE REVIEW

**Comment by Trancever** [comment]
Hi @joaom182 How about passing `KeyboardAvoidingView` as a child to `Modal` by yourself? E.g:
```
<Modal>
  <KeyboardAvoidingView>
    other components
  </KeyboardAvoidingView>
</Modal>
```

Our library should not be responsible for handling keyboard and to be honest it's very unlikely that we will merge this PR.

**Comment by HarisSpahija** [comment]
> Hi @joaom182 How about passing `KeyboardAvoidingView` as a child to `Modal` by yourself? E.g:
> 
> ```
> <Modal>
>   <KeyboardAvoidingView>
>     other components
>   </KeyboardAvoidingView>
> </Modal>
> ```
> Our library should not be responsible for handling keyboard and to be honest it's very unlikely that we will merge this PR.

The thing is that the dialog is even breaking on a smaller screens where menu's are slided up. The dialog should be responsive. I dont see a breaking change with this PR, its a nice QOL patch.

**Comment by Trancever** [comment]
@HarisSpahija Could you provide [Snack](https://snack.expo.io/) with example of Modal/Dialog that breaks?

**Comment by HarisSpahija** [comment]
@Trancever I think OP posted an example here: [OP example](https://github.com/callstack/react-native-paper/issues/957#issuecomment-477997386)

I will try to reproduce the bug

**Comment by HarisSpahija** [comment]
@Trancever On IOS. Create any dialog in React-Native. Open keyboard while the dialog is open (Default `Command ⌘ + K`). 
Result: Keyboard is overlapping the dialog.

File: src/components/Modal.js [hunk 1]
```diff
@@ -165,6 +167,7 @@ class Modal extends React.Component<Props, State> {
 
     const { children, dismissable, theme, contentContainerStyle } = this.props;
     const { colors } = theme;
+    const SurfaceWrapper = Platform.OS === 'ios' ? KeyboardAvoidingView : View;
```
Comment (ferrannp) [inline src/components/Modal.js]: We need to make this optional and make it default. Can you pass a prop and change base branch to 3.0?

Comment (joaom182) [reply src/components/Modal.js]: Can u suggest the prop name?

Comment (Trancever) [reply src/components/Modal.js]: @joaom182 It can be keyboardAware of boolean type. Also I would prefer to keep existing behaviour as default.

**Comment by fawaz-ahmed-toptal** [comment]
Wrap `<Dialog>` inside the following component `KeyboardAwareDialog`, like this :

```
<KeyboardAwareDialog
          dismissable={false}
          visible={this.state.visible}
        >
       <Dialog.Title>
         {'Over the keyboard'}
       </Dialog.Title>
</KeyboardAwareDialog
```
here is the file `KeyboardAwareDialog.js`:
```
import React, { memo, useState, useEffect } from 'react';
import { View, Dimensions } from 'react-native';
import { Dialog, Colors } from 'react-native-paper';
import KeyboardSpacer from 'react-native-keyboard-spacer';

const windowHeight = Dimensions.get('window').height;
const halfWindowHeight = windowHeight / 2;

export default memo(({ children, ...rest }) => {
  const [topSpacing, setTopSpacing] = useState(10);
  const [height, setHeight] = useState(0);

  const onLayout = ({nativeEvent: { layout: {height : _height}}}) => {
    if (!height && height !== _height) {
      setHeight(_height);
    }
  };

  useEffect(() => {
    const newTopSpacing = - halfWindowHeight + height;
    setTopSpacing(newTopSpacing);
  }, [height])

  return (
    <Dialog
      {...rest}
      style={{ backgroundColor: 'transparent' }}
    >
      <View onLayout={onLayout} style={{ backgroundColor: Colors.white }}>
        {children}
      </View>
      <KeyboardSpacer topSpacing={topSpacing} />
    </Dialog>
  );
});
```

**Comment by wardjk87** [comment]
Since it's not needed for android I use Ternary: 
`  const DialogType = Platform.OS === 'ios' ? KeyboardAwareDialog : Dialog;`

**Comment by ersumanta** [comment]
> Wrap `<Dialog>` inside the following component `KeyboardAwareDialog`, like this :
> 
> ```
> <KeyboardAwareDialog
>           dismissable={false}
>           visible={this.state.visible}
>         >
>        <Dialog.Title>
>          {'Over the keyboard'}
>        </Dialog.Title>
> </KeyboardAwareDialog
> ```
> 
> here is the file `KeyboardAwareDialog.js`:
> 
> ```
> import React, { memo, useState, useEffect } from 'react';
> import { View, Dimensions } from 'react-native';
> import { Dialog, Colors } from 'react-native-paper';
> import KeyboardSpacer from 'react-native-keyboard-spacer';
> 
> const windowHeight = Dimensions.get('window').height;
> const halfWindowHeight = windowHeight / 2;
> 
> export default memo(({ children, ...rest }) => {
>   const [topSpacing, setTopSpacing] = useState(10);
>   const [height, setHeight] = useState(0);
> 
>   const onLayout = ({nativeEvent: { layout: {height : _height}}}) => {
>     if (!height && height !== _height) {
>       setHeight(_height);
>     }
>   };
> 
>   useEffect(() => {
>     const newTopSpacing = - halfWindowHeight + height;
>     setTopSpacing(newTopSpacing);
>   }, [height])
> 
>   return (
>     <Dialog
>       {...rest}
>       style={{ backgroundColor: 'transparent' }}
>     >
>       <View onLayout={onLayout} style={{ backgroundColor: Colors.white }}>
>         {children}
>       </View>
>       <KeyboardSpacer topSpacing={topSpacing} />
>     </Dialog>
>   );
> });
> ```

Fantastic solution. worked first time.

**Comment by papuruth** [comment]
> Wrap `<Dialog>` inside the following component `KeyboardAwareDialog`, like this :
> 
> ```
> <KeyboardAwareDialog
>           dismissable={false}
>           visible={this.state.visible}
>         >
>        <Dialog.Title>
>          {'Over the keyboard'}
>        </Dialog.Title>
> </KeyboardAwareDialog
> ```
> 
> here is the file `KeyboardAwareDialog.js`:
> 
> ```
> import React, { memo, useState, useEffect } from 'react';
> import { View, Dimensions } from 'react-native';
> import { Dialog, Colors } from 'react-native-paper';
> import KeyboardSpacer from 'react-native-keyboard-spacer';
> 
> const windowHeight = Dimensions.get('window').height;
> const halfWindowHeight = windowHeight / 2;
> 
> export default memo(({ children, ...rest }) => {
>   const [topSpacing, setTopSpacing] = useState(10);
>   const [height, setHeight] = useState(0);
> 
>   const onLayout = ({nativeEvent: { layout: {height : _height}}}) => {
>     if (!height && height !== _height) {
>       setHeight(_height);
>     }
>   };
> 
>   useEffect(() => {
>     const newTopSpacing = - halfWindowHeight + height;
>     setTopSpacing(newTopSpacing);
>   }, [height])
> 
>   return (
>     <Dialog
>       {...rest}
>       style={{ backgroundColor: 'transparent' }}
>     >
>       <View onLayout={onLayout} style={{ backgroundColor: Colors.white }}>
>         {children}
>       </View>
>       <KeyboardSpacer topSpacing={topSpacing} />
>     </Dialog>
>   );
> });
> ```

This doesn't work when android:windowSoftInputMode is set to "adjustPan" and when set to "adjustResize" the Dialog moves extremely up and there is lots of space between keyboard and Dialog.

**Pull request closed without merging**